#   Question 1 

#    Random Forest Company Data 

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
from sklearn import preprocessing 
from sklearn.preprocessing import StandardScaler 
from sklearn.model_selection import train_test_split
from sklearn import tree
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report

In [2]:
company_data = pd.read_csv("Company_Data.csv")
company_data

Sales  CompPrice  Income  Advertising  Population  Price ShelveLoc  Age  \
0     9.50        138      73           11         276    120       Bad   42   
1    11.22        111      48           16         260     83      Good   65   
2    10.06        113      35           10         269     80    Medium   59   
3     7.40        117     100            4         466     97    Medium   55   
4     4.15        141      64            3         340    128       Bad   38   
..     ...        ...     ...          ...         ...    ...       ...  ...   
395  12.57        138     108           17         203    128      Good   33   
396   6.14        139      23            3          37    120    Medium   55   
397   7.41        162      26           12         368    159    Medium   40   
398   5.94        100      79            7         284     95       Bad   50   
399   9.71        134      37            0          27    120      Good   49   

     Education Urban   US  
0           17   Yes  Yes  
1           10   Yes  Yes  
2           12   Yes  Yes  
3           14   Yes  Yes  
4           13   Yes   No  
..         ...   ...  ...  
395         14   Yes  Yes  
396         11    No  Yes  
397         18   Yes  Yes  
398         12   Yes  Yes  
399         16   Yes  Yes  

[400 rows x 11 columns]

In [3]:
company_data.describe()

Sales   CompPrice      Income  Advertising  Population  \
count  400.000000  400.000000  400.000000   400.000000  400.000000   
mean     7.496325  124.975000   68.657500     6.635000  264.840000   
std      2.824115   15.334512   27.986037     6.650364  147.376436   
min      0.000000   77.000000   21.000000     0.000000   10.000000   
25%      5.390000  115.000000   42.750000     0.000000  139.000000   
50%      7.490000  125.000000   69.000000     5.000000  272.000000   
75%      9.320000  135.000000   91.000000    12.000000  398.500000   
max     16.270000  175.000000  120.000000    29.000000  509.000000   

            Price         Age   Education  
count  400.000000  400.000000  400.000000  
mean   115.795000   53.322500   13.900000  
std     23.676664   16.200297    2.620528  
min     24.000000   25.000000   10.000000  
25%    100.000000   39.750000   12.000000  
50%    117.000000   54.500000   14.000000  
75%    131.000000   66.000000   16.000000  
max    191.000000   80.000000   18.000000

###  Preprocessing the dataset

In [4]:
company_data['Sales'] = pd.cut(company_data.Sales, bins=(-1.0,7.49,16.27),
                               labels=['Low','High'])

In [5]:
company_data

Sales  CompPrice  Income  Advertising  Population  Price ShelveLoc  Age  \
0    High        138      73           11         276    120       Bad   42   
1    High        111      48           16         260     83      Good   65   
2    High        113      35           10         269     80    Medium   59   
3     Low        117     100            4         466     97    Medium   55   
4     Low        141      64            3         340    128       Bad   38   
..    ...        ...     ...          ...         ...    ...       ...  ...   
395  High        138     108           17         203    128      Good   33   
396   Low        139      23            3          37    120    Medium   55   
397   Low        162      26           12         368    159    Medium   40   
398   Low        100      79            7         284     95       Bad   50   
399  High        134      37            0          27    120      Good   49   

     Education Urban   US  
0           17   Yes  Yes  
1           10   Yes  Yes  
2           12   Yes  Yes  
3           14   Yes  Yes  
4           13   Yes   No  
..         ...   ...  ...  
395         14   Yes  Yes  
396         11    No  Yes  
397         18   Yes  Yes  
398         12   Yes  Yes  
399         16   Yes  Yes  

[400 rows x 11 columns]

In [6]:
company_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   Sales        400 non-null    category
 1   CompPrice    400 non-null    int64   
 2   Income       400 non-null    int64   
 3   Advertising  400 non-null    int64   
 4   Population   400 non-null    int64   
 5   Price        400 non-null    int64   
 6   ShelveLoc    400 non-null    object  
 7   Age          400 non-null    int64   
 8   Education    400 non-null    int64   
 9   Urban        400 non-null    object  
 10  US           400 non-null    object  
dtypes: category(1), int64(7), object(3)
memory usage: 31.9+ KB


In [7]:
label_encoder = preprocessing.LabelEncoder()
company_data["Sales"] = label_encoder.fit_transform(company_data["Sales"])
company_data["ShelveLoc"] = label_encoder.fit_transform(company_data["ShelveLoc"])
company_data["Urban"] = label_encoder.fit_transform(company_data["Urban"])
company_data["US"] = label_encoder.fit_transform(company_data["US"])

In [8]:
company_data

Sales  CompPrice  Income  Advertising  Population  Price  ShelveLoc  Age  \
0        0        138      73           11         276    120          0   42   
1        0        111      48           16         260     83          1   65   
2        0        113      35           10         269     80          2   59   
3        1        117     100            4         466     97          2   55   
4        1        141      64            3         340    128          0   38   
..     ...        ...     ...          ...         ...    ...        ...  ...   
395      0        138     108           17         203    128          1   33   
396      1        139      23            3          37    120          2   55   
397      1        162      26           12         368    159          2   40   
398      1        100      79            7         284     95          0   50   
399      0        134      37            0          27    120          1   49   

     Education  Urban  US  
0           17      1   1  
1           10      1   1  
2           12      1   1  
3           14      1   1  
4           13      1   0  
..         ...    ...  ..  
395         14      1   1  
396         11      0   1  
397         18      1   1  
398         12      1   1  
399         16      1   1  

[400 rows x 11 columns]

###  Splitting the dataset

In [9]:
x = company_data.iloc[:,1:11]
y = company_data.iloc[:,0]

In [10]:
X_train, X_test, y_train, y_test = train_test_split(x,y, test_size=0.3, random_state=0)

###   Standardizing the dataset

In [11]:
x_train = StandardScaler().fit_transform(X_train)
x_test = StandardScaler().fit_transform(X_test)

###   Building Random Forest 

In [12]:
model = RandomForestClassifier(n_estimators=200, criterion='entropy',random_state=0)
model.fit(x_train,y_train)

RandomForestClassifier(criterion='entropy', n_estimators=200, random_state=0)

###  Predicting the model 

In [13]:
y_pred = model.predict(x_test)
y_pred

array([1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1,
       0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0,
       1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0,
       0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1,
       0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1,
       0, 1, 0, 0, 1, 1, 1, 0, 1, 1])

###  Evaluating the model 

In [14]:
confusion_matrix(y_test, y_pred)

array([[50, 15],
       [ 6, 49]], dtype=int64)

In [15]:
print("Accuracy:", accuracy_score(y_test, y_pred))

Accuracy: 0.825


In [16]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.89      0.77      0.83        65
           1       0.77      0.89      0.82        55

    accuracy                           0.82       120
   macro avg       0.83      0.83      0.82       120
weighted avg       0.83      0.82      0.83       120



###  Feature importance 

In [17]:
feature_importances_df = pd.DataFrame(
    {"feature": list(x.columns), "importance": model.feature_importances_}
).sort_values("importance", ascending=False)
feature_importances_df

feature  importance
4        Price    0.252591
0    CompPrice    0.137196
6          Age    0.126934
5    ShelveLoc    0.108095
1       Income    0.096436
2  Advertising    0.095182
3   Population    0.094235
7    Education    0.054923
8        Urban    0.018275
9           US    0.016135

### Using random forest the accuracy of model is improved to 82%